ARROW-7725: [C++] Add infrastructure for unity builds and precompiled headers

pitrou opened the pull request:
Unity builds and precompiled headers can be enabled "easily" in CMake 3.16:
https://cmake.org/cmake/help/v3.16/prop_tgt/UNITY_BUILD.html
https://cmake.org/cmake/help/v3.16/command/target_precompile_headers.html

They can make builds faster in some conditions, especially on CI with little parallelism and caching.

Some care is required to make them work:

- unity builds concatenate a number of C/C++ source files together,
  so there should be no symbol clashes, C++ namespace resolution issues
  or conflicting preprocessor definitions

- precompiled headers require that the same compile options are passed
  for all files using the header

pitrou commented:
Example GHA normal build:
https://github.com/apache/arrow/runs/417128878
Same build in unity mode => 5 minutes faster:
https://github.com/pitrou/arrow/runs/417253236

kszucs commented:
> Example GHA normal build:
> https://github.com/apache/arrow/runs/417128878
> Same build in unity mode => 5 minutes faster:
Nice!
> https://github.com/pitrou/arrow/runs/417253236

kou commented on cpp/cmake_modules/BuildUtils.cmake:
```diff
@@ -173,21 +191,26 @@ function(ADD_ARROW_LIB LIB_NAME)
     if(ARG_DEPENDENCIES)
       add_dependencies(${LIB_NAME}_objlib ${ARG_DEPENDENCIES})
     endif()
+    if(ARG_PRECOMPILED_HEADER_LIB AND ARROW_USE_PRECOMPILED_HEADERS)
```
Could you remove `AND ARROW_USE_PRECOMPILED_HEADERS` because this is checked in `reuse_precompiled_header_lib`?

pitrou replied:
Will do.

kou commented on cpp/src/arrow/dataset/pch.h:
```diff
@@ -0,0 +1,25 @@
+// before checking in.  Adding headers which are not used extremely often
+// may incur a slowdown, since it makes the precompiled header heavier to load.
+
+#include "arrow/dataset/dataset.h"
```
(This is not related to this change.)

If we use `#include "..."`, we should use relative path of the current directory:

```cpp
#include "dataset.h"
#include "scanner.h"
#include "../pch.h"
```

If we don't use relative path, we should use `#include <...>` style.

See also:
* https://gcc.gnu.org/onlinedocs/cpp/Include-Syntax.html
* https://docs.microsoft.com/en-us/cpp/preprocessor/hash-include-directive-c-cpp?view=vs-2019

pitrou commented:
I removed unused stuff and addressed your review comments @kou.

A small benchmark, clang 9.0, ccache disabled, Ubuntu 18.04, debug build:
```
# !UNITY & !PCH
real	1m7,932s
user	12m14,896s
sys	0m35,349s

# !UNITY & PCH
real	0m59,173s
user	10m39,391s
sys	0m33,052s

# UNITY & !PCH
real	0m49,501s
user	8m22,581s
sys	0m22,592s

# UNITY & PCH
real	0m48,917s
user	7m34,525s
sys	0m21,088s
```

pitrou commented:
AppVeyor build here: https://ci.appveyor.com/project/pitrou/arrow/builds/30748930

@kou Do you want to review this again?

kou commented on cpp/src/parquet/column_reader.cc:
```diff
@@ -110,7 +110,7 @@ ReaderProperties default_reader_properties() {
 // assembled in a serialized stream for storing in a Parquet files
 
-static constexpr int16_t kNonPageOrdinal = static_cast<int16_t>(-1);
+// static constexpr int16_t kNonPageOrdinal = static_cast<int16_t>(-1);
```
Can we remove this?

pitrou replied:
Oops, forgot it. Thank you.

kou commented:
Umm. Failed again.

Does this error relate to this change?

https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/30818324/job/ilu5bu0422bn2s5q#L1207

```text
C:/projects/arrow/cpp/src/parquet/arrow/schema.cc(265): error C2039: 'DecimalSize': is not a member of 'parquet::arrow::internal'
C:/projects/arrow/cpp/src/parquet/arrow/reader.cc(928): note: see declaration of 'parquet::arrow::internal'
C:/projects/arrow/cpp/src/parquet/arrow/schema.cc(265): error C3861: 'DecimalSize': identifier not found
```

pitrou commented:
> Does this error relate to this change?

Yes, it does. As mentioned in the PR message, unity builds can lead to C++ namespace resolution issues. Let's say you have two different files:

```c++
// a.cpp

namespace arrow {
namespace io {
namespace internal {

void SomeFunction();
}
}
}
```

```c++
// b.cpp

namespace arrow {
namespace io {
using internal::OtherFunction;
}
}
```

If `b.cpp` is compiled separately, `internal::OtherFunction` can refer to `::arrow::internal::OtherFunction`. But if `a.cpp` and `b.cpp` are concatenated together, then `internal::OtherFunction` will refer to `::arrow::io::internal::OtherFunction` which may not exist.

kou commented:
+1
